### PR TITLE
EventFilter add filter for ConfigMap Endpoints

### PR DESCRIPTION
**Pull request by yanfeng1992**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:

Some components on the old k8s version still use configmap and endpoints for leader election, I think these resources should be ignored

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
`karmada-controller-manager`:  EventFilter add filter for ConfigMap Endpoints
```

### CODE REVIEW

**Comment by XiShanYongYe-Chang** [comment]
Hi @yanfeng1992 thanks a lot~
We will be releasing a new version (v1.7.0) by the end of this month (8.31), ref to #3997 
Let's continue this PR when the next version starts.

File: pkg/detector/detector.go [hunk 1]
```diff
@@ -276,6 +277,14 @@ func (d *ResourceDetector) EventFilter(obj interface{}) bool {
 			if found && secretType == string(corev1.SecretTypeServiceAccountToken) {
 				return false
 			}
```
Comment (chaunceyjiang) [inline pkg/detector/detector.go:279]: Can you add some comments to here?

Comment (yanfeng1992) [reply pkg/detector/detector.go:280]: // The configmap and endpoints for leader election should be prevented from propagating. 

@chaunceyjiang

**Review by chaunceyjiang — COMMENTED** [review]
LGTM

File: pkg/detector/detector.go [hunk 2]
```diff
@@ -271,11 +272,20 @@ func (d *ResourceDetector) EventFilter(obj interface{}) bool {
...
 			secretType, found, _ := unstructured.NestedString(unstructObj.Object, "type")
 			if found && secretType == string(corev1.SecretTypeServiceAccountToken) {
 				return false
 			}
+		case corev1.SchemeGroupVersion.WithKind("ConfigMap"):
+			if _, ok := unstructObj.GetAnnotations()[resourcelock.LeaderElectionRecordAnnotationKey]; ok {
+				return false
+			}
+		case corev1.SchemeGroupVersion.WithKind("Endpoints"):
+			if _, ok := unstructObj.GetAnnotations()[resourcelock.LeaderElectionRecordAnnotationKey]; ok {
```
Comment (jwcesign) [inline pkg/detector/detector.go:286]: How about `Lease` resource?

**Review by jwcesign — COMMENTED** [review]
I find there are many other builtin resources, for example:
```
-> # ka get secret -A
NAMESPACE         NAME                   TYPE     DATA   AGE
karmada-cluster   member1                Opaque   2      27h
karmada-cluster   member1-impersonator   Opaque   1      27h
karmada-cluster   member2                Opaque   2      27h
karmada-cluster   member2-impersonator   Opaque   1      27h
karmada-cluster   member3                Opaque   2      27h
karmada-cluster   member3-impersonator   Opaque   1      27h
```

But there isn't a good way to do it.
/cc @XiShanYongYe-Chang @whitewindmills @chaunceyjiang 

What do you think?

File: pkg/detector/detector.go (hunk 2 above)
Comment (yanfeng1992) [reply pkg/detector/detector.go:286]: https://github.com/karmada-io/karmada/blob/7c96e0db541989bd95772bbc1cc0109d8479371c/pkg/util/apigroup.go#L54C3-L54C3

NewSkippedResourceConfig  disable Lease by default

Comment (jwcesign) [reply pkg/detector/detector.go:286]: Thanks for reminding me.

**Comment by whitewindmills** [comment]
@jwcesign 
it's okay, they will be ignored.
https://github.com/karmada-io/karmada/blob/7c96e0db541989bd95772bbc1cc0109d8479371c/pkg/detector/detector.go#L236-L246

**Review by whitewindmills — COMMENTED** [review]
good work. 👍 
/lgtm

File: pkg/detector/detector.go [hunk 3]
```diff
@@ -271,11 +272,20 @@ func (d *ResourceDetector) EventFilter(obj interface{}) bool {
 		// The secret, with type 'kubernetes.io/service-account-token', is created along with `ServiceAccount` should be
 		// prevented from propagating.
 		// Refer to https://github.com/karmada-io/karmada/pull/1525#issuecomment-1091030659 for more details.
+		// The configmap and endpoints for leader election should be prevented from propagating.
```
Comment (XiShanYongYe-Chang) [inline pkg/detector/detector.go:275]: Shouldn't this comment be placed below? Other LGTM

Comment (yanfeng1992) [reply pkg/detector/detector.go]: ok，thanks for reminding

**Review by RainbowMango — COMMENTED** [review]
/assign

**Comment by XiShanYongYe-Chang** [comment]
Thanks
/lgtm

**Comment by RainbowMango** [comment]
> Some components on the old k8s version still use configmap and endpoints for leader election,

Do you know from which version Kubernetes stopped using ConfigMap and Endpoints and started using Lease?

My concern is if the version prior to 1.16, we might not need to ignore them as Karmada apiserver requires kube-apiserver >= 1.16 (Karmada's CRD is generated with CRD/v1 (Kubernetes 1.16)).

**Comment by yanfeng1992** [comment]
leaderelection in client-go supports 5 types of resource locks:

const (
    LeaderElectionRecordAnnotationKey = "[control-plane.alpha.kubernetes.io/leader](http://control-plane.alpha.kubernetes.io/leader)"
    EndpointsResourceLock             = "endpoints"
    ConfigMapsResourceLock            = "configmaps"
    LeasesResourceLock                = "leases"
    EndpointsLeasesResourceLock       = "endpointsleases"
    ConfigMapsLeasesResourceLock      = "configmapsleases"
)

In version 1.24, the community completely removed support for Endpoints/ConfigMaps as the primary resource lock.

https://github.com/kubernetes/kubernetes/pull/106852

https://github.com/kubernetes/kubernetes/issues/80289

These two types(endpointsleases, configmapsleases) are in a transition period, and it is necessary to obtain both Endpoints/Configmap and Lease resource locks (MultiLock) at the same time. The loss of any resource lock will cause the component to restart. When updating the lock, update Endpoints/Configmap first, and then update Lease. When judging loss, if both resources have holders but are inconsistent, an exception will be returned and the owner will be re-elected.

These two types of MultiLock still exist in our actual environment(v1.25.8). Of course, the overall trend in the community is to use the Leases type instead of the original Endpoints/Configmaps as resource locks. But we need to consider supporting MultiLock in transition phase

image

image

@RainbowMango

**Comment by RainbowMango** [comment]
> These two types of MultiLock still exist in our actual environment(v1.25.8). 

If I remember right, all Karmada components now use `Lease` as locks, so, my question is why the other lock resources would exist in your environment?

Here is the lease I just get from `karmada-apiserver`:
```
# kubectl get leases.coordination.k8s.io -A 
NAMESPACE         NAME                         HOLDER                                                                                 AGE
karmada-cluster   member3                      member3                                                                                20h
karmada-system    karmada-agent-member3        karmada-agent-76cbbbc75-6djbm_c52b1d77-e152-4e56-8669-f62786a96116                     20h
karmada-system    karmada-controller-manager   karmada-controller-manager-5d8cc94745-67j85_88e0201c-a9f7-4834-bee6-5fab52e3e633       20h
karmada-system    karmada-descheduler          karmada-descheduler-7d654cbff6-sztzd_cfae23f7-ea84-429c-b81e-b4e6ac719d2c              20h
karmada-system    karmada-scheduler            karmada-scheduler-58b454779b-gj6gs_1e2a71e9-0e32-41f7-b73e-cbe183f990d2                20h
kube-system       kube-controller-manager      karmada-kube-controller-manager-66866b6d4-t6hgd_85a868e2-7eac-4a82-b954-1c3433947113   20h

```

**Comment by yanfeng1992** [comment]
In our multi-cluster model, there are some other business team controllers that need resources in watch karmada-apiserver

image

**Comment by RainbowMango** [comment]
> In our multi-cluster model, there are some other business team controllers that need resources to watch karmada-apiserver

Now I get it.  The business controllers use the `configmaps` or `endpoints` as the lock. 

One possible approach is just like this PR, let Karmada ignore these potential resources, but before that, I'm thinking a question: what's the problem here if Karmada does not do it?
If I understand correctly, without the event filter, these resources will be put on the `waiting list` and be there forever. No functionality issues, but they will be processed (non-opt) periodicity because the business controller would update them.

Another possible approach is to let the business controller start to use `Lease`, is it possible for you?

In addition, this case is quite similar to #3866, which requests to define an allow-list to control what kind of resource should be watched by the detector.

**Comment by yanfeng1992** [comment]
> Another possible approach is to let the business controller start to use `Lease`, is it possible for you?

yes i agree with that，use `Lease` is a better way

>  No functionality issues, but they will be processed (non-opt) periodicity because the business controller would update them.

If the update frequency is relatively high, can it also be considered a problem? What I see now is that it is updated every two seconds, which is also related to the configuration of the main election

**Comment by RainbowMango** [comment]
> If the update frequency is relatively high, can it also be considered a problem? What I see now is that it is updated every two seconds, which is also related to the configuration of the main election

I don't think you need to renew the lock too often, for Karmada, take `karmada-controller-manager` as an example,  the renew duration is `10s`.
```
      --leader-elect-renew-deadline duration                                                                                                                                                        
                The interval between attempts by the acting master to renew a leadership slot before it stops leading. This must be less than or equal to the lease duration. This is only applicable
                if leader election is enabled. (default 10s)
```

In addition to the `configmaps` and `endpoints`, there must be some other resources we need to filter, what I'm thinking is how to provide an extensible or configurable mechanism to handle these things instead of hard-code them one by one.

**Comment by yanfeng1992** [comment]
> In addition to the `configmaps` and `endpoints`, there must be some other resources we need to filter, what I'm thinking is how to provide an extensible or configurable mechanism to handle these things instead of hard-code them one by one.

Okay, I agree with this too

**Pull request closed without merging**